src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/resource/LicenceController.kt

```diff
@@ -0,0 +1,62 @@
+package uk.gov.justice.digital.hmpps.createandvaryalicenceapi.resource
+
+import io.swagger.v3.oas.annotations.Operation
+import io.swagger.v3.oas.annotations.responses.ApiResponse
+import io.swagger.v3.oas.annotations.responses.ApiResponses
+import io.swagger.v3.oas.annotations.security.SecurityRequirement
+import org.springframework.http.MediaType
+import org.springframework.security.access.prepost.PreAuthorize
+import org.springframework.web.bind.annotation.GetMapping
+import org.springframework.web.bind.annotation.PathVariable
+import org.springframework.web.bind.annotation.PostMapping
+import org.springframework.web.bind.annotation.RequestBody
+import org.springframework.web.bind.annotation.RequestMapping
+import org.springframework.web.bind.annotation.ResponseBody
+import org.springframework.web.bind.annotation.RestController
+import uk.gov.justice.digital.hmpps.createandvaryalicenceapi.model.CreateLicenceRequest
+import uk.gov.justice.digital.hmpps.createandvaryalicenceapi.model.CreateLicenceResponse
+import uk.gov.justice.digital.hmpps.createandvaryalicenceapi.model.Licence
+import uk.gov.justice.digital.hmpps.createandvaryalicenceapi.service.LicenceService
+import javax.validation.constraints.NotEmpty
+
+@RestController
+@RequestMapping("/licence", produces = [MediaType.APPLICATION_JSON_VALUE])
+class LicenceController(private val licenceService: LicenceService) {
+
+  @PostMapping(value = ["/create"])
+  @PreAuthorize("hasAnyRole('SYSTEM_USER', 'CVL_ADMIN')")
+  @ResponseBody
+  @Operation(
+    summary = "Create a licence",
+    description = "Creates a licence with the default status IN_PROGRESS and populates it with the details provided.",
+    security = [SecurityRequirement(name = "ROLE_SYSTEM_USER"), SecurityRequirement(name = "ROLE_CVL_ADMIN")],
+  )
+  @ApiResponses(
+    value = [
+      ApiResponse(responseCode = "401", description = "Unauthorised, requires a valid Oauth2 token"),
+      ApiResponse(responseCode = "403", description = "Forbidden, requires an appropriate role")
+    ]
+  )
+  fun createLicence(@RequestBody @NotEmpty request: CreateLicenceRequest): CreateLicenceResponse {
+    return licenceService.createLicence(request)
+  }
+
+  @GetMapping(value = ["/id/{licenceId}"])
+  @PreAuthorize("hasAnyRole('SYSTEM_USER', 'CVL_ADMIN')")
+  @ResponseBody
+  @Operation(
+    summary = "Get a licence by its licence id",
+    description = "Returns a single licence detail by its unique identifier.",
+    security = [SecurityRequirement(name = "ROLE_SYSTEM_USER"), SecurityRequirement(name = "ROLE_CVL_ADMIN")],
+  )
+  @ApiResponses(
+    value = [
+      ApiResponse(responseCode = "401", description = "Unauthorised, requires a valid Oauth2 token"),
```

Comment:
hmm, instead of defining these for each endpoint, can we consider creating a controller advice class from spring and annotating controllers with @ControllerAdvice, it means we can just throw exceptions if an error happens and the controller will be advised as to which API response to return. We can create custom exceptions as well to create further flexibility. But this can be another task

Reply:
We already have a ControllerAdvice class for catching exceptions and creating an Error response. These values are really to show in the swagger docs what sorts of errors can be expected and what they mean.

Reply:
Cool thanks